### PR TITLE
Array ufunc remove dot matmul

### PR DESCRIPTION
**Pull request by mhvk**
(see main thread)

### CODE REVIEW

**Comment by charris** [comment]
Not really reversion. In git parlance a reversion undoes a previous commit.

**Comment by mhvk** [comment]
Feel free to change the commit titles accordingly... (I can do it too if you wish)

**Comment by charris** [comment]
I'm waiting for @pv to weigh in on this as `dot` was in the original for a reason and I'd like to know what it was.

**Comment by pv** [comment]
The aim was to make it possible to write type generic code also for matrix multiplication.

**Comment by mhvk** [comment]
That why I'd really hope to have `matmul` covered, but via a gufunc....  I think we should simply assume this will happen, as in outline it is known what to do. And then we can still follow @njsmith's suggestion of pushing the most common cases for `np.dot` through the same ufunc.

**Comment by pv** [comment]
The situation currently is of course a bit different with the `@` notation available.
However, `__array_ufunc__` is currently irrelevant for scipy.sparse in its current form because we'll have to set `__array_ufunc__ = None` there.

**Comment by charris** [comment]
So would it be reasonable for the `@` operator to behave like the other binary operators, i.e., check `__array_priority__` and `__array_ufunc__` and return `NotImplemented` if indicated? Note that that doesn't require that the actual implementation use the override mechanism.

**Comment by pv** [comment]
That `@` behaves like other binary operations vs. ufunc opt-out probably would be the expected behavior.

**Comment by mhvk** [comment]
> However, `__array_ufunc__` is currently irrelevant for scipy.sparse in its current form because we'll have to set `__array_ufunc__ = None` there.

Just to be clear, this has to be done to be able to continue to override `ndarray.__mul__`, ie., being able to distinguish between `array * sparse ` and `np.multiply(array, sparse)`?

In an earlier incarnation, @njsmith had special-cased `scipy` but in the end, I think that is not the right way to proceed. But might there be a way for `scipy.sparse` itself to determine how it was called? (e.g., the python stack would show that things went through `operator` rather than directly into the ufunc?).

Anyway, more directly to the point of this sub-PR: does it make sense to ensure that `__array_ufunc__` actually gets passed in a `ufunc`, i.e., remove `np.dot` and `np.matmul` until they are ufuncs?

**Comment by pv** [comment]
Correct (ditto for addition of sage matrices). Realistically, stacktracing will not be done because it's too ugly. I doubt it can be done at all, because the reason `np.multiply` was entered inside another class's `__mul__` probably cannot be inferred.

**Comment by mhvk** [comment]
OK, makes sense. In any case, this is really a question of whether we'd allow ufuncs and binops to do something different, and I think the consensus was that we'd at the very least not encourage that in any way. (If we want to reconsider that, we probably better discuss in the main thread; this one really should be about whether I did the override removal of `np.dot` and `np.matmul` correctly).

File: numpy/core/src/multiarray/multiarraymodule.c [hunk 1]
```diff
@@ -2365,47 +2342,33 @@ array_matmul(PyObject *NPY_UNUSED(m), PyObject *args, PyObject* kwds)
...
-                "Cannot find a common data type.");
+        PyErr_SetString(PyExc_ValueError, "Cannot find a common data type.");
         return NULL;
     }
     typenum = dtype->type_num;
 
     if (typenum == NPY_OBJECT) {
         /* matmul is not currently implemented for object arrays */
         PyErr_SetString(PyExc_TypeError,
-                "Object arrays are not currently supported");
```
Comment (charris) [inline numpy/core/src/multiarray/multiarraymodule.c]: Conventionally we break the style guidelines for the error cases in order to allow longer messages.

Comment (mhvk) [inline numpy/core/src/multiarray/multiarraymodule.c]: But in this case the message fits within 80 characters.  Should one still break the style guide in that case?

Comment (charris) [inline numpy/core/src/multiarray/multiarraymodule.c]: Not worth worrying about. I probably should have said "**_I_** break the style..." ;)

**Pull request merged**